### DISCUSSION TITLE
Trigger submit on change

### DISCUSSION THREAD

**Discussion by marcobiedermann** [react-hook-form/react-hook-form#8486] [Ideas]
## Description

It would be great to have an option to call `handleSubmit` on change instead of form submission.
This would allow the form to be submitted if any of the fields change e.g. auto-save/auto-search.

There is already an option for validation but not for submitting the form.

```ts
useForm({
  mode: 'onChange'
})
```

Listening on change is currently only possible via `watch`

```ts
useEffect(() => {
  const formData = watch()

  console.log({ formData })
}, [watch()])
```

## Proposal

```ts
useForm({
  submitMode: 'onChange'
})
```

## References

Formik auto-save: https://github.com/jaredpalmer/formik/issues/1218#issuecomment-466927384

**Comment by bluebill1049** [comment]
could you do something like this?

```tsx
useEffect(() => {
  watch((dta) => {
    // you can throttle the request here as well, and without any render to the UI
    saveData(data)
  })
}, [watch])
```

**Reply by andrevinsky** [reply]
@bluebill1049 kindly asking to please explain your reluctance to include such option in the library itself. I.e. the file's `onChange` is one of the cases where the functionality to submit will come right off handy. Not naming other reasons I came across while searching for solution to my task.

**Reply by marcobiedermann** [reply]
@bluebill1049 works, TypeScript is just complaining about some types

**Reply by k-zakhariy** [reply]
https://codesandbox.io/s/react-hook-form-auto-save-xgulp?file=/src/AutoSave.tsx

**Reply by wj127** [reply]
As of today no official answer...it would have been very nice seeing this feature added into the library

**Reply by xkomiks** [reply]
I agree that this is a necessary feature in such a popular library.

**Comment by florinkrasniqi** [comment]
After trying out @k-zakhariy's method, which worked but for my case didn't see the need of having a custom file for that, I just tweaked the code a bit and used it in one of my form like this:


```typescript
import { useDeepCompareEffect } from 'use-deep-compare';
import debounce from 'lodash.debounce';

const debouncedSave = useCallback(
  debounce(() => {
    form.handleSubmit(props.onSubmit)();
  }, 1000),
  [form, props.onSubmit],
);

const watchedData = form.watch();

useDeepCompareEffect(() => {
  if (form.formState.isDirty) {
    debouncedSave();
  }
}, [watchedData]);
```

**Comment by DedaDev** [comment]
With debounce, if anyone needs it:

```ts

    const debouncedSubmit = useMemo(
        () => debounce((data: FilterForm) => {
            submitForm(data);
        }, 500),
        [submitForm]
    );

    useEffect(() => {
        const subscription = watch((data) => {
            debouncedSubmit(data as FilterForm);
        });

        return () => {
            subscription.unsubscribe();
            debouncedSubmit.cancel();
        };
    }, [watch, debouncedSubmit]);


```
